### ISSUE TITLE
[Optimisations] [X86] Storing an `i4 zeroext` into an `i8` then loading that `i8` still creates code to store the `i8` even when not used

### ISSUE THREAD

**Issue by MitalAshok** [llvm/llvm-project#97224]
C code:

```c
bool f(const unsigned _BitInt(4) x) {
  return __builtin_bit_cast(unsigned char, x) == 3;
}
```

Output https://godbolt.org/z/xPjWsYYPj: (x86-64)

```asm
f:                                      # @f
        mov     byte ptr [rsp - 1], dil
        cmp     dil, 3
        sete    al
        ret
```

With an unused store on the stack.

And since the parameter is an `i4 noundef zeroext` (i.e., it has 4 padding bits that are all 0), it should also be able to optimize:

```c
bool f(const unsigned _BitInt(4) x) {
  return __builtin_bit_cast(unsigned char, x) == 0b01110000;
}
```

... to become `return false`, but it becomes:

```asm
f:                                      # @f
        mov     byte ptr [rsp - 1], dil
        cmp     dil, 112
        sete    al
        ret
```

**Comment by e-kud** [comment]
It has been fixed somewhere between 18.1.0 and 19.1.0
Current generation is
```
f:
        cmp     edi, 3 // no extra load
        sete    al
        ret

f1:
        xor     eax, eax // optimized out
        ret
```
Probably we can close it.

**Comment by nikic** [comment]
Yes, Clang no longer generates non-byte-sized accesses for BitInt.
